### PR TITLE
Add configuration option for LDAP referral

### PR DESCRIPTION
**Pull request by tsaarni**
Allows setting LDAP referral property to:
- ignore (default)
- follow

This PR adds support in LDAP federation provider to set `java.naming.referral` property and a new UI option to configure it. 

Since [documentation](https://www.keycloak.org/docs/latest/server_admin/#other-configuration-options) does not list individual LDAP configuration settings, I did not add referrals to docs either.

Updates #13500

### CODE REVIEW

File: js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx [hunk 1]
```diff
@@ -403,6 +404,38 @@ export const LdapSettingsSearching = ({
...
         </FormGroup>
+        <FormGroup
+          label={t("referral")}
+          labelIcon={
+            <HelpItem helpText={t("referralHelp")} fieldLabelId="referral" />
+          }
+          fieldId="kc-referral"
+        >
+          <Controller
+            name="config.referral[0]"
```
Comment (jonkoops) [inline js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx:415]: I do believe newer versions of React Hook Form prefer the following syntax:
```suggestion
            name="config.referral.0"
```

File: js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx [hunk 2]
```diff
@@ -403,6 +404,38 @@ export const LdapSettingsSearching = ({
...
+            defaultValue=""
+            control={form.control}
+            render={({ field }) => (
+              <Select
+                toggleId="kc-referral"
+                required
```
Comment (jonkoops) [inline js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx:421]: Is this a required field? There are no annotations on the `FormField` to indicate so.
```suggestion
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx (hunk 2 above)
Comment (tsaarni) [reply js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx]: No, it is not required. I removed `required`.

Not sure what is the practice, but I did not add empty select option for the menu. Currently the selection defaults to `ignore` and that reflects how it works on the Java side - the JDK LDAP implementation uses `ignore` by default, even when the `java.naming.referral` property is not set at all.

File: js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx (hunk 1 above)
Comment (tsaarni) [reply js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx]: Done!

File: js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx (hunk 2 above)
Comment (jonkoops) [reply js/apps/admin-ui/src/user-federation/ldap/LdapSettingsSearching.tsx]: That's probably fine, I will consider this resolved.

**Review by jonkoops — DISMISSED** [review]
Can't vouch for the Java stuff, but the client-side code LGTM.

File: js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties [hunk 3]
```diff
@@ -2916,3 +2916,5 @@ invalidEmailMessage='{{0}}': Invalid email address.
 missingLastNameMessage='{{0}}': Please specify last name.
 missingEmailMessage='{{0}}': Please specify email.
 missingPasswordMessage='{{0}}': Please specify password.
+referral=Referral
+referralHelp=Specifies if LDAP referrals should be followed or not.
```
Comment (ahus1) [inline js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties:2920]: Nitpick:

```suggestion
referralHelp=Specifies if LDAP referrals should be followed or ignored.
```

**Review by ahus1 — CHANGES_REQUESTED** [review]
"Not followed" is actually two-fold in Java: to ignore them or to throw an exception. See the Javadoc for `Context.REFERAL` - see below for a change I'd like to see.

Do you want to add the third option "throw" as well?

I wonder if this option should come with a warning: Referrals might be slow, and you might end up contacting a server which you didn't plan to contact / don't trust. Please give it a thought and let me know if you would put such a warning/hint into the description. I don't want to prevent Keycloak from using those referrals as some (corporate) users might need them to work in their environment, but some explanations might be good.

**Comment by tsaarni** [comment]
> "Not followed" is actually two-fold in Java: to ignore them or to throw an exception. See the Javadoc for `Context.REFERAL` - see below for a change I'd like to see.
> 
> Do you want to add the third option "throw" as well?

Since we do not specifically have any functionality associated to `ReferralException` I was thinking I would not add `throw` option. The only use case I could imagine, is that we'd instruct user to set `throw` and then go check the logs to learn if referral was received or not. But this seems clumsy use case to me...  On the other hand, if you think this could be useful feature, I have nothing against adding `throw` option.

By the way, there is also undocumented [`follow-scheme`](https://github.com/openjdk/jdk17u/blob/40c29f7d80b5b67718ce071722bd7cfaecdf9495/src/java.naming/share/classes/com/sun/jndi/ldap/LdapCtx.java#L2445) which will follow only if the scheme in referral was the same as in original URI. I did not add this since I did not find it documented anywhere.

> I wonder if this option should come with a warning: Referrals might be slow, and you might end up contacting a server which you didn't plan to contact / don't trust. Please give it a thought and let me know if you would put such a warning/hint into the description. I don't want to prevent Keycloak from using those referrals as some (corporate) users might need them to work in their environment, but some explanations might be good.

Good point! I added a note for that. Please check and tell me what you think.

**Review by ahus1 — APPROVED** [review]
Thank you for this change!

**Review by jonkoops — APPROVED** [review]

**Comment by tsaarni** [comment]
Thank you @jonkoops and @ahus1!

**Pull request merged**